### PR TITLE
Refactor field resolution closer to schema types

### PR DESCRIPTION
**Pull request by dackroyd**
Much of the work of executing resolvers relates to the resolvable portion of fields, rather than the selection made for the request. With the introduction of directive visitors, these were being recreated for each request, instead of being built once, and re-used.

By pushing the resolution of fields down into the `resolvable.Field`, directives applied to the field definition can be setup just once when the schema is created, and re-used across all requests. Small exceptions to this exist for field types that embed this to provide request specific values (resolver instance and arguments), which get passed along with the call.

The chain of directive visitor functions still needs to be recreated for each resolve call, because the inner resolve function needs to accept the request specific field resolver instance, but other per-request overheads should be removed.

### CODE REVIEW

File: internal/exec/resolvable/resolvable.go [hunk 1]
```diff
@@ -158,6 +226,44 @@ func ApplyResolver(s *types.Schema, resolver interface{}, dirVisitors []directiv
...
+	builder := packer.NewBuilder()
+
+	packers := map[string]*packer.StructPacker{}
+	for _, d := range s.Directives {
+		n := d.Name
+
+		v, ok := visitors[n]
+		if !ok {
+			// Directives which need visitors have already been checked
+			// Anything without a visitor now is an in-built directive without a packer.
```
Comment (pavelnikolov) [inline internal/exec/resolvable/resolvable.go:241]: It might be a directive that is just "decorative". I would reword:
```suggestion
			// Directives which need visitors have already been checked.
			// Anything without a visitor now is a directive which doesn't need a packer.
```

File: internal/exec/selected/selected.go [hunk 2]
```diff
@@ -153,21 +162,14 @@ func applySelectionSet(r *Request, s *resolvable.Schema, e *resolvable.Object, s
...
-					PackedArgs:       packedArgs,
-					PackedDirectives: packedDirectives,
-					Sels:             fieldSels,
-					Async:            fe.HasContext || fe.ArgsPacker != nil || fe.HasError || HasAsyncSel(fieldSels),
+					Field:      *fe,
+					Alias:      field.Alias.Name,
+					Args:       args,
+					PackedArgs: packedArgs,
+					Sels:       fieldSels,
+					Async:      fe.HasContext || fe.ArgsPacker != nil || fe.HasError || HasAsyncSel(fieldSels),
```
Comment (pavelnikolov) [inline internal/exec/selected/selected.go:172]: Should fields with directive packers be async too? I guess yes... 
This operation takes a bit longer than a normal field resolution. 🤔 
I think now is the time to add one more condition:
```suggestion
					Async:      fe.HasContext || fe.ArgsPacker != nil || fe.DirectiveVisitors != nil || fe.HasError || HasAsyncSel(fieldSels),
```

**Review by pavelnikolov — CHANGES_REQUESTED** [review]
Thank you, Dave! I left a few comments but other than that it looks good to me.

File: internal/exec/resolvable/resolvable.go [hunk 3]
```diff
@@ -158,6 +226,44 @@ func ApplyResolver(s *types.Schema, resolver interface{}, dirVisitors []directiv
...
+			continue
+		}
+
+		if _, ok := v.(directives.ResolverInterceptor); !ok {
```
[the remark this answers is not in the document]
Comment (pavelnikolov) [reply internal/exec/resolvable/resolvable.go:245]: Apparently, when you start a draft PR, once you click "Ready for review" it bypasses checks by SemaphoreCI 🤦‍♂️ 
This marks master as broken because of [the same issue last time](https://graph-gophers.semaphoreci.com/jobs/83bb4db9-e210-49d3-a1e4-df0a7e31f4ef). I'll see how I can fix it.

File: internal/exec/resolvable/resolvable.go [hunk 4]
```diff
@@ -47,7 +47,7 @@ type Field struct {
 	HasContext        bool
 	HasError          bool
 	ArgsPacker        *packer.StructPacker
-	DirectivesPackers map[string]*packer.StructPacker
+	DirectiveVisitors []directives.ResolverInterceptor
```
[the remark this answers is not in the document]
Comment (dackroyd) [reply internal/exec/resolvable/resolvable.go:50]: Yep its something that needs to change. In changes I'm making on top of this to add validators, I've had to replace this so that both kinds of directive handling can be retrieved with the right subset for each place they need to run

File: internal/exec/resolvable/resolvable.go [hunk 5]
```diff
@@ -56,6 +56,69 @@ func (f *Field) UseMethodResolver() bool {
...
+	}
+
+	for _, pd := range f.DirectiveVisitors {
+		pd := pd // Needed to avoid passing only the last directive, since we're closing over this loop var pointer
+		innerResolver := currResolver
+
+		currResolver = func(ctx context.Context, args interface{}) (output interface{}, err error) {
+			return pd.Resolve(ctx, args, resolverFunc(innerResolver))
+		}
+	}
```
[the remark this answers is not in the document]
Comment (dackroyd) [reply internal/exec/resolvable/resolvable.go:77]: I suspect at some stage `d` was already used while moving this around, so had to use something else (and the types were renamed a few times). I'll revisit
